osquery/tables/events/freebsd/openbsm_events.cpp

```diff
@@ -0,0 +1,208 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+#include <arpa/inet.h>
+#include <netinet/in.h>
+#include <sys/socket.h>
+
+#include <bsm/libbsm.h>
+
+#include <osquery/events.h>
+#include <osquery/logger.h>
+
+#include "osquery/events/darwin/openbsm.h"
+
+namespace osquery {
+
+namespace tables {
+extern long getUptime();
+}
+
+static inline void OpenBSM_AUT_SUBJECT32_EX(Row& r, const tokenstr_t& tok) {
+  if (tok.id != AUT_SUBJECT32_EX) {
+    return;
+  }
+  r["auid"] = INTEGER(tok.tt.subj32_ex.auid);
+  r["pid"] = INTEGER(tok.tt.subj32_ex.pid);
+  r["uid"] = INTEGER(tok.tt.subj32_ex.ruid);
+  r["gid"] = INTEGER(tok.tt.subj32_ex.rgid);
+  r["euid"] = INTEGER(tok.tt.subj32_ex.euid);
+  r["egid"] = INTEGER(tok.tt.subj32_ex.egid);
+  char ip_str[INET6_ADDRSTRLEN];
+  if (tok.tt.proc32_ex.tid.type == AU_IPv4) {
+    struct in_addr ipv4;
+    ipv4.s_addr = static_cast<in_addr_t>(*tok.tt.subj32_ex.tid.addr);
+    r["address"] =
+        std::string(inet_ntop(AF_INET, &ipv4, ip_str, INET6_ADDRSTRLEN));
+  } else {
+    struct in6_addr ipv6;
+    memcpy(&ipv6, tok.tt.subj32_ex.tid.addr, sizeof(ipv6));
+    r["address"] =
+        std::string(inet_ntop(AF_INET6, &ipv6, ip_str, INET6_ADDRSTRLEN));
+  }
+}
+
+static inline unsigned long decimalIntToOctInt(unsigned long x) {
+  auto ret = 0;
+  for (auto i = 1; x > 0; i *= 10) {
+    ret += (x & 0x7) * i;
+    x >>= 3;
+  }
+  return ret;
+}
+
+class OpenBSMExecVESubscriber : public EventSubscriber<OpenBSMEventPublisher> {
+ public:
+  Status init() override {
+    return Status(0);
+  }
+
+  void configure() override;
+
+  Status Callback(const OpenBSMEventContextRef& ec,
+                  const OpenBSMSubscriptionContextRef& sc);
+};
+
+class OpenBSMSSHLoginSubscriber
+    : public EventSubscriber<OpenBSMEventPublisher> {
+ public:
+  Status init() override {
+    return Status(0);
+  }
+
+  void configure() override;
+
+  Status Callback(const OpenBSMEventContextRef& ec,
+                  const OpenBSMSubscriptionContextRef& sc);
+};
+
+REGISTER(OpenBSMExecVESubscriber, "event_subscriber", "process_events");
+REGISTER(OpenBSMSSHLoginSubscriber, "event_subscriber", "user_events");
+
+void OpenBSMExecVESubscriber::configure() {
+  std::vector<size_t> event_ids{
```

Comment:
Are there other ids that should be added here that are specific to FreeBSD?

Reply:
Hm, perhaps fexecve (43144)? And maybe 43131, which is a MAC-related version of it. (MAC as opposed to DAC).